File: 2493-reverse-odd-levels-of-binary-tree/2493-reverse-odd-levels-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private Queue<TreeNode> reverse(Queue<TreeNode> q){
        List<TreeNode> l= new ArrayList<>(q);
        int start=0,end=q.size()-1;
        while(start<end){
            int temp=l.get(start).val;
            l.get(start).val=l.get(end).val;
            l.get(end).val=temp;
            start++;
            end--;
        }
        Queue<TreeNode> newQ=new LinkedList<>(l);
        return newQ;
    }
    public TreeNode reverseOddLevels(TreeNode root) {
        Queue<TreeNode> q=new LinkedList<>();
        q.offer(root);
        int lvl=0;
        while(!q.isEmpty()){
            int l=q.size();
            if(lvl%2==1){
                q=reverse(q);
            }
            for(int i=0;i<l;i++){
                TreeNode r=q.poll();
                if(r.left!=null){
                    q.offer(r.left);
                }
                if(r.right!=null){
                    q.offer(r.right);
                }
            }
            lvl++;
        }
        return root;
    }
}

File: 2493-reverse-odd-levels-of-binary-tree/2493-reverse-odd-levels-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def reverseOddLevels(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def reverse(q):
            start,end=0,len(q)-1
            while start<end:
                q[start].val,q[end].val=q[end].val,q[start].val
                start+=1
                end-=1
            return q
        def printd(lvl,q):
            print(lvl,':',end=' ')
            for i in q:
                print(i.val,end=' ')
            print('\n------------------------')
        q=deque([root])
        lvl=0
        while q:
            l=len(q)
            printd(lvl,q)
            if lvl%2==1:
                q=reverse(q)
            for i in range(l):
                r=q.popleft()
                if r.left:
                    q.append(r.left)
                if r.right:
                    q.append(r.right)
            lvl+=1
        return root